### PR TITLE
[v2] Make tree-sitter optional, add convenience functions

### PR DESCRIPTION
**Pull request by jvdydev**
Addresses #366.
Additionally addresses undocumented issue where tree-sitter requires a C compiler to compile grammars (somewhat hinted in #363).

Now, configuring tree-sitter works as follows:
```emacs-lisp
(require 'crafted-ide-packages)

(package-install-selected-packages :noconfirm)

(require 'crafted-ide-config)

;; Emacs 28 or earlier
(crafted-ide-configure-tree-sitter)

;; Emacs 29 or later
;; To achieve the same behaviour as before
(crafted-ide-configure-tree-sitter)

;; Optionally pass a list of opt-out language grammars
(crafted-ide-configure-tree-sitter '(protobuf)) ; don't set up protobuf grammar
```

Documentation got two new sections (one for 28 and ealier, one for 29 and later).
Regenerated info buffer.

Note that this is a "fix" for these issues, although I'm not happy about the fact that language grammars are opt-in for pre-29 and opt-out for post-29. It should be either opt-in for both or neither.

### CODE REVIEW

**Comment by sthesing** [comment]
> Now, configuring tree-sitter works as follows:
> 
> ```emacs-lisp
> (require 'crafted-ide-packages)
> 
> (package-install-selected-packages :noconfirm)
> 
> (require 'crafted-ide-config)
> 
> ;; Emacs 28 or earlier
> (crafted-ide-configure-tree-sitter)
> 
> ;; Emacs 29 or later
> ;; To achieve the same behaviour as before
> (crafted-ide-configure-tree-sitter)
> 
> ;; Optionally pass a list of opt-out language grammars
> (crafted-ide-configure-tree-sitter '(protobuf)) ; don't set up protobuf grammar
> ```

I haven't had time to look at the code, yet. But just looking at the result,  #366 asked for an easy way to opt out of using tree-sitter. Now, it is opt-in for everyone. I think I'd prefer it to be an opt-out model.

> Note that this is a "fix" for these issues, although I'm not happy about the fact that language grammars are opt-in for pre-29 and opt-out for post-29. It should be either opt-in for both or neither.

Hm... I haven't really looked into the topic of language grammar installation, but it seems to me that the main problem to for making this opt-in or opt-put for everyone, regardless of the version they're using, is that there is no definite list of grammars that would need to be turned off or on in the respective other version. We could solve this by defining a list of "supported" language grammars. And that, I think, makes more sense for an opt-out model.

**Comment by jvdydev** [comment]
ye, the reason I went for opt-in is it allows people to use `crafted-ide` on platforms without (easy) access to a C compiler (locked-down installations, Android, ...).
We had a similar discussion on `org-roam` a few months ago since it requires a locally compiled SQLite database (and we also don't auto-install `pdf-tools` for similar reasons).

I understand tree-sitter is *very nice* to have, but I don't think it's a good idea to enable it by default if it requires people to set up external things outside of Emacs (personally).

The only "opt-out" mechanism that would actually satisfy #366 would be a `defcustom` wrapping `treesit-auto-opt-out-list` (e.g. `crafted-ide-treesit-opt-out-list`):
```emacs-lisp
(require 'crafted-ide-packages)

(package-install-selected-packages :noconfirm)

(customize-set-variable 'crafted-ide-treesit-opt-out-list '(protobuf))
(require 'crafted-ide-config)
```
This is because as shown in #366, `treesit-auto-opt-out-list` has to be modified in-between requiring `treesit-auto` and calling `treesit-auto-install-all`.

File: modules/crafted-ide-config.el [hunk 1]
```diff
@@ -48,27 +48,32 @@ manually with something like this:
 (customize-set-variable 'eglot-autoshutdown t)
 
 ;;; tree-sitter
-;; Emacs versions prior to 29
-(when (version< emacs-version "29")
-  (when (require 'tree-sitter-indent nil :noerror)
+(defun crafted-ide--configure-tree-sitter-pre-29 ()
+  "Configure tree-sitter for Emacs 28 or earlier."
+  (when (version< emacs-version "29")
+    (when (require 'tree-sitter-indent nil :noerror)
```
Comment (jeffbowman) [inline modules/crafted-ide-config.el:54]: Why is this here?  The code which follows doesn't seem to need `tree-sitter-indent` unless I'm missing something - which is possible.

File: docs/crafted-ide.org [hunk 2]
```diff
@@ -42,7 +42,46 @@ operating system. See [[https://github.com/joaotavora/eglot#connecting-to-a-serv
...
+
+(crafted-tree-sitter-load 'python)
+
+#+end_src
+
+**** Configuring Tree-Sitter (Emacs 29 or later)
+
+Call ~crafted-ide-configure-tree-sitter~ after requiring ~crafted-ide-config~.
+This will install all known language grammars for Tree-Sitter.
+To opt-out of one or more language grammars, pass them as a list
```
Comment (jeffbowman) [inline docs/crafted-ide.org:71]: Specifically mentioning the protobuf looping problem from issue #366 might be a useful example of when this might be applicable.

**Review by jeffbowman — CHANGES_REQUESTED** [review]

File: modules/crafted-ide-config.el (hunk 1 above)
Comment (jvdydev) [reply modules/crafted-ide-config.el:54]: Well, the version guard I just forgot to remove (as it's handled by `crafted-ide-configure-tree-sitter`).
As for the require guard, that was already there in the previous version.
I'll remove them both.

File: docs/crafted-ide.org (hunk 2 above)
Comment (jvdydev) [reply docs/crafted-ide.org:71]: yes, I'll also add to the docs that using tree-sitter requires a C compiler to compile the language grammars.

**Review by jeffbowman — APPROVED** [review]

File: modules/crafted-ide-config.el (hunk 1 above)
Comment (jeffbowman) [reply modules/crafted-ide-config.el]: D'oh!  Probably my fault then.  I wonder why I did that.  Thanks for fixing it!
